test/util/lifecycle.js

```diff
@@ -15,15 +15,16 @@ import createTestContainer from './container';
 from the vm property of the Vue Test Utils wrapper: see
 https://github.com/vuejs/test-utils/issues/2140. In practice, the only global
 properties that we try to access in testing are $router and $route. We can
-provide access to them by shadowing the global properties with computed
-properties from a mixin. */
+provide access to them by copying $router to each component, then computing
+$route based on $router. */
 const shadowRouterProps = {
+  created() {
+    // eslint-disable-next-line no-self-assign
+    if (this.$router != null) this.$router = this.$router;
+  },
   computed: {
-    $router() {
-      return this.$container.router;
-    },
     $route() {
-      return this.$router.currentRoute.value;
+      return this.$router != null ? this.$router.currentRoute.value : undefined;
     }
```

Comment:
We don't assign `$route` to `this` like we do with `$router` because the value of `$route` can change (whenever there's a navigation and the route changes). Unlike `$router`, `$route` as a global property seems to be a getter: https://github.com/vuejs/router/blob/a1611b6099403cfe9539ee8d4e9308b3a6a0175c/packages/router/src/router.ts#L1211-L1215